src/ConnectionAutomaton.h

```diff
@@ -50,7 +50,8 @@ class ConnectionAutomaton :
     /// Registered as an input in the DuplexConnection.
     public Subscriber<std::unique_ptr<folly::IOBuf>>,
     /// Receives signals about connection writability.
-    public Subscription {
+    public Subscription,
+    public std::enable_shared_from_this<ConnectionAutomaton> {
```

Comment:
YMMV, but I've tried using `enable_shared_from_this` in the past, and it seems _very_ hard to get right. In particular, the restriction that there must already be an `std::shared_ptr` to `this` is difficult to satisfy in all cases—and getting it wrong is undefined behavior (not necessarily a crash).

If you do want to pursue this approach, I would suggest making all constructors `private` and exposing a `static` method which returns the `std::shared_ptr` correctly.